onnxruntime/core/providers/webgpu/tensor/scatter_nd.cc

```diff
@@ -0,0 +1,263 @@
+        break;
+      default:
+        ORT_THROW("Unsupported reduction type: ", static_cast<int>(reduction));
+        // The controlflow should never reach here.
+    }
+    return ss.str();
+  };
+
+  auto calc_data_offset_snippet = [](uint32_t output_rank, bool parallel) -> std::string {
+    std::ostringstream ss;
+    // The stride is size is one less than the shape size.
+    if (output_rank == 1) {
+      ss << "    let element_count_dim = 1u;\n";
+    } else if (output_rank == 2) {
+      ss << "    let element_count_dim = select(uniforms.output_stride, 1u, " << (parallel ? "i - indices_start" : "i") << " == 1 " << ");\n";
+    } else {
+      ss << "    let output_stride_index = " << (parallel ? "i - indices_start" : "i") << ";\n"
+         << "    let element_count_dim = select(uniforms.output_stride[output_stride_index], 1u, output_stride_index == " << (output_rank - 1) << ");\n";
+    }
+    if (output_rank == 1) {
+      ss << "    let dim_value = uniforms.output_shape;\n";
+    } else {
+      ss << "    let dim_value = uniforms.output_shape[" << (parallel ? "i - indices_start" : "i") << " + uniforms.last_index_dimension];\n";
+    }
+
+    ss << "    if (index >= 0) {\n"
+       << "      if (index >= i32(dim_value)) {\n"
+       << "        index = i32(dim_value - 1);\n"
+       << "      }\n"
+       << "    } else {\n"
+       << "      if (index < -i32(dim_value)) {\n"
+       << "        index = 0;\n"
+       << "      } else {\n"
+       << "        index += i32(dim_value);\n"
+       << "      }\n"
+       << "    }\n"
+       << "    data_offset += u32((u32(index) * element_count_dim));\n";
+    return ss.str();
+  };
+
+  auto update_elements_snippet = [atomic_reduction_snippet](ScatterNDReduction reduction, const std::string& data_type, bool parallel) -> std::string {
+    std::ostringstream ss;
+    ss << "  for (var i = 0u; i < uniforms.num_updates_elements; i++) {\n"
+       << "    let value = updates[uniforms.num_updates_elements * " << (parallel ? "global_idx" : "idx") << " + i];\n"
+       << atomic_reduction_snippet(reduction, "output[data_offset + i]", "value", data_type) << "\n"
+       << "  }\n";
+    return ss.str();
+  };
+  std::string data_type_str;
+  bool reducible = false;
+  if (data_type_ == DataTypeImpl::GetType<int32_t>()) {
+    reducible = true;
+    data_type_str = "i32";
+  } else if (data_type_ == DataTypeImpl::GetType<uint32_t>()) {
+    reducible = true;
+    data_type_str = "u32";
+  } else if (data_type_ == DataTypeImpl::GetType<float>()) {
+    reducible = true;
+    data_type_str = "f32";
+  } else {
+    // Default value.
+    data_type_str = "output_element_t";
+  }
+  if (reduction_ != ScatterNDReduction::None && !reducible) {
+    ORT_THROW("ScatterND: Reduction is not supported for data type ", data_type_str);
+  }
+  shader.MainFunctionBody() << shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.output_size")
+                            << "  var hasDuplicates = false;\n";
```

Comment:
maybe this should be shared memory so we don't need to always do this?

Reply:
I will remove the code computing ` hasDuplicates` as suggested.